File: src/jss2013/G.cpp
#include <iostream>
#include <algorithm>
using namespace std;

#define INFINITE 100000
int leftmintime[1001];
int rightmintime[1001];
int N,X,Y,MAX;
struct Board
{
    int lx;
    int rx;
    int high;
}board[1001];

bool cmp(Board a,Board b)
{
    if(a.high > b.high) return true;
    else return false;
}
int mintime(int index,bool left)
{
    int i,x;
    int y = board[index].high;
    if(left)
        x = board[index].lx;
    else
        x = board[index].rx;
    for(i = index + 1;i <= N;i++)
        if(board[i].lx <= x && board[i].rx >= x)
            break;
    
    if(i <= N){ // 板index的下方有板
        if(y - board[i].high > MAX)
            return INFINITE;
    }
    else{
        if(y > MAX)
            return INFINITE;
        else
            return y;
    }
    int lefttime = y - board[i].high + x - board[i].lx;
    int righttime = y - board[i].high + board[i].rx - x;
    if(leftmintime[i] == -1)
        leftmintime[i] = mintime(i,true);
    if(rightmintime[i] == -1)
        rightmintime[i] = mintime(i,false);
    lefttime+=leftmintime[i];
    righttime+=rightmintime[i];
    if(lefttime < righttime)
        return lefttime;
    else
        return righttime;
}

int main()
{
    int t;
    
    cin >> t;
    while(t--){ 
        int i;
        memset(leftmintime,-1,sizeof(leftmintime));
        memset(rightmintime,-1,sizeof(rightmintime));
        cin >> N >> X >> Y >> MAX;
        for(i = 1;i <= N;i++){
            cin >> board[i].lx >> board[i].rx >> board[i].high;
        }
        sort(board,board+N,cmp);  //对板从上到下一次排序编号

        board[0].lx = X;
        board[0].rx = X;
        board[0].high = Y;
        cout << mintime(0,true);
    }
    return 0;
}

File: src/jss2013/G.java
package jss2013;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;
import java.util.Arrays;

public class G {
	static StreamTokenizer st;
	static int n, x, y, max;
	static Board[] board = new Board[1001];
	static int[] lt = new int[1001];
	static int[] rt = new int[1001];
	public static void main(String[] args) throws IOException {
		st = new StreamTokenizer(new BufferedReader(new InputStreamReader(System.in)));
		int t = nextInt();
		while(t-->0){
			n = nextInt();
			x = nextInt();
			y = nextInt();
			max = nextInt();
			
			Arrays.fill(lt, -1);
			Arrays.fill(rt,  -1);
			for(int i=1;i<=n;i++){
				int lx = nextInt();
				int rx = nextInt();
				int y = nextInt();
				board[i] = new Board(lx, rx, y);
			}
			
			board[0] = new Board(x, x, y);
			Arrays.sort(board, 0, n);
			
			System.out.println(mintime(0, true));
		}
	}
	
	static int mintime(int idx, boolean left){
		int x,y;
		if(left)
			x = board[idx].lx;
		else
			x = board[idx].rx;
		y = board[idx].y;
		int i;
		for(i=idx+1;i<=n;i++)
			if(board[i].lx <= x && board[i].rx >= x)
				break;
		if(i>n){
			if(y > max)
				return Integer.MAX_VALUE;
			else
				return y;
		}
		if(y - board[i].y > max)
			return Integer.MAX_VALUE;
		
		if(lt[i] < 0)
			lt[i] = mintime(i, true);
		if(rt[i] < 0)
			rt[i] = mintime(i, false);
		
		int fall = y - board[i].y;
		int lefttime = fall + x - board[i].lx + lt[i];
		int righttime = fall + board[i].rx - x + rt[i];
		
		return Math.min(lefttime, righttime);
	}
	
	static int nextInt() throws IOException{
		st.nextToken();
		return (int)st.nval;
	}
	
}

class Board implements Comparable<Board>{
	int lx, rx, y;
	Board(int l, int r, int h){
		lx = l;
		rx = r;
		y = h;
	}
	@Override
	public int compareTo(Board arg0) {
		return arg0.y - this.y;
	}
}